packages/strapi-plugin-upload/admin/src/containers/InputModalStepperProvider/reducer.js

```diff
@@ -0,0 +1,169 @@
+import produce from 'immer';
+import { intersectionWith, differenceWith, unionWith } from 'lodash';
+
+import createNewFilesToUploadArray from '../../utils/createNewFilesToUploadArray';
+
+const initialState = {
+  selectedFiles: [],
+  files: [],
+  filesToUpload: [],
+  fileToEdit: null,
+  tab: null,
+  params: {
+    _limit: 10,
+    _start: 0,
+    _q: '',
+    filters: [],
+  },
+  currentStep: 'list',
+};
+
+const reducer = (state, action) =>
+  // eslint-disable-next-line consistent-return
+  produce(state, draftState => {
+    switch (action.type) {
+      case 'ON_CHANGE': {
+        draftState.fileToEdit[action.keys.split('.')] = action.value;
+        break;
+      }
+      case 'GET_DATA_COUNT_SUCCEEDED': {
+        draftState.count = action.data.count;
+        break;
+      }
+      case 'GET_DATA_SUCCEEDED': {
+        draftState.files = action.data;
+        break;
+      }
+      case 'SET_PARAM': {
+        const { name, value } = action.param;
+
+        if (name === 'filters') {
+          draftState.params.filters.push(value);
+          break;
+        }
+
+        if (name === '_limit') {
+          draftState.params._start = 0;
+        }
+
+        draftState.params[name] = value;
+        break;
+      }
+      case 'ON_FILE_SELECTION': {
+        const { value, id } = action;
+
+        if (value) {
+          const fileToStore = state.files.find(file => file.id === id);
+          draftState.selectedFiles.push(fileToStore);
+          break;
+        }
+
+        const index = draftState.selectedFiles.findIndex(item => item.id === id);
+        draftState.selectedFiles.splice(index, 1);
+        break;
+      }
+      case 'TOGGLE_SELECT_ALL': {
+        const comparator = (first, second) => first.id === second.id;
+        const isSelected =
+          intersectionWith(state.selectedFiles, state.files, comparator).length ===
+          state.files.length;
+
+        if (isSelected) {
+          draftState.selectedFiles = differenceWith(state.selectedFiles, state.files, comparator);
```

Comment:
I think you can simplify this logic if I understand it you are trying to avoid to have duplicate files so maybe you can do this by using a simple filter. I find this code hard to understand.

Reply:
The difference function of lodash will just return the difference between two arrays. All arrays manipulation can be done with Array prototype methods but as we have lodash, can't we just use it? What do you think?